### DISCUSSION TITLE
Cannot serve pages after install - Multiple questions for setup/HA

### DISCUSSION THREAD

**Discussion by di-rect** [osm-search/Nominatim#2169] [General]
I'm following the following guide for installing Nominatim, but some questions raise.

[https://nominatim.org/release-docs/latest/appendix/Install-on-Ubuntu-20/](url)

After a build I'm not able to serve the pages, Nginx tells me the following:

`2021/02/15 12:08:30 [error] 20#20: *1 FastCGI sent in stderr: "Unable to open primary script: \/opt/nominatim/build/website\/.php (No such file or directory)" while reading response header from upstream, client: 192.168.250.1, server: , request: "GET / HTTP/1.1", upstream: "fastcgi://unix:/var/run/php8.0-fpm.sock:", host: "192.168.250.201:88"`

- I have Git cloned all contents into /opt/nominatim
- Created build in /opt/nominatim and build from there

I also wonder what files are relevant only after build, is this only everything in /opt/nominatim/build or more ?

I would like to put PHP-FPM on a seperate server but I'm quite sure I run into the issue that both Nginx and PHP-FPM need to have the sources then this shared by NFS or so, am I right or is there another way to accomplish that ?

I also wonder what the best way is to connect from Nominatim to an external PostgreSQL server, where/how to set ? I have read that the PosgreSQL server needs to have the nominatim.so available on it's install ?

What is the idea about this all ? I would like to have a HA Nominatim setup with shared data so I can balance there as well.

Ideas are welcome!

**Comment by mtmail** [comment]
Note the https://nominatim.org/release-docs/latest/ is for the stable Nominatim v3.6 release (which is probably want you want for a production system). After 3.6 there have been big changes to script names, paths, configuration file formats and more changes are coming.

I can't comment on php-fpm on remote host, maybe a user on the https://lists.openstreetmap.org/listinfo/geocoding has done this before.

> I also wonder what files are relevant only after build, is this only everything in /opt/nominatim/build or more ?

In version 3.6 scripts in build/ are tiny wrappers. E.g. a build/utils/warm.php will call Nominatim/utils/warm.php and use Nominatim/lib/ PHP files. So you will need the path you cloned (or rather the files from the stable *.tar.gz release) plus everything in the build directory.

> Nominatim to an external PostgreSQL server,

Base configuration is `build/settings/settings.php`, which gets overloaded by `build/settings/local.php`. In there look for `CONST_Database_DSN`. Format is https://www.php.net/manual/en/ref.pdo-pgsql.connection.php

The `nominatim.so` file needs to be copied to the database server. Or compiled on the database server. See https://nominatim.org/release-docs/latest/admin/Advanced-Installations/#importing-nominatim-to-an-external-postgresql-database

**Reply by di-rect** [reply]
Clear. I was git cloning the release branch so I assume that is the same one ?

Checking out the nginx config I run into some things I cannot find anywhere:

If I comment:

```
#        if (!-f \$document_root\$fastcgi_script_name) {
#            return 404;
#        }
```

Then the PHP page get served but I get other errors I don't know why they exist:

```
Warning: require_once(/opt/nominatim/build/settings/settings-frontend.php): Failed to open stream: No such file or directory in /opt/nominatim/build/website/search.php on line 3

Fatal error: Uncaught Error: Failed opening required '/opt/nominatim/build/settings/settings-frontend.php' (include_path='.:/usr/share/php') in /opt/nominatim/build/website/search.php:3 Stack trace: #0 {main} thrown in /opt/nominatim/build/website/search.php on line 3 
```

Is the Nginx config really working/up-to-date ? I see various versions in Docker builds and most of the time Apache being used.

I will test out the other options, also moving the nominati.so after build to PgSQL. PHP tear out is what I will test. It must be doable but need the /build mount in the same wat so I'm not sure if it does matter to do so.

**Reply by mtmail** [reply]
I don't see a release branch named 3.6 on https://github.com/osm-search/Nominatim/branches/stale I assume you meant the git tag or commit hash mentioned on https://github.com/osm-search/Nominatim/releases/tag/v3.6.0

https://nominatim.org/release-docs/3.6.0/admin/Import/#setting-up-the-website describes how the `settings-frontend.php` gets created.

Most users use the easiest, Apache2, webserver but we know people also use php-fpm. I use it and didn't see the errors you describe.

**Reply by di-rect** [reply]
Thanks for the quick reply and helping out.

Yes I use 3.6.0 for the branch, the same version as for the download. Should be the same I believe, download will be generated from branch I assume.

Why do the docs say the API is just there after the steps which are described ? It seems you need ./utils/setup.php --setup-website which might be a typo there then ?

https://nominatim.org/release-docs/3.6.0/appendix/Install-on-Ubuntu-20/#option-2-using-nginx

FPM is just nice so I can scale it, I use it a lot and it doesn't make any difference to sock at some point, yes it makes difference with the pools. Wel designed you can scale very fast so better to start with it then need to refactor by default.

You use the same Nginx config or do you use Apache ? I wouldn't blame you for it ;)

**Reply by lonvia** [reply]
The missing `./utils/setup.php --setup-website` was an oversight. It's a new thing to do in 3.6 and we forgot to adapt the documentation. Fixed now.

**Reply by lonvia** [reply]
Regarding serving methods: apache is nice and easy to set up for those with little Linux system administrator experience. If you know your way around PHP-FPM and nginx already, then by all means use it. Apache has known scaling issues that can cause severe head aches.

If you want to scale the setup, then you should be aware that the database is the bottle neck and the point where you need to multiply resources. For 3.6., just keep the whole source and build directory around on each machine that has webservice (i.e. PHP-FPM) running and also on each machine where you run database updates. This will all become a bit easier once we have `make install` in the next release.
